rfcs/0075-declarative-wrappers.md

```diff
@@ -0,0 +1,367 @@
+both ideas could be integrated into an alternative, simpler creation method of
+binary wrappers. See [my
+comment](https://github.com/NixOS/nixpkgs/pull/95569#issuecomment-674508806).
+
+- `hardware.sane` module installs sane impurly
+- [issue 90201](https://github.com/NixOS/nixpkgs/issues/90201)
+- [issue 90184](https://github.com/NixOS/nixpkgs/issues/90184)
+
+The current way NixOS enables to configure access to scanner, is via the
+`hardware.sane` module, which interacts [just a
+bit](https://github.com/NixOS/nixpkgs/blob/5d8dd5c2598a74761411bc9bef7c9111d43d2429/nixos/modules/services/hardware/sane.nix#L34)
+wish `saned`, but mostly does nothing besides setting `SANE_CONFIG_DIR` and
+`LD_LIBRARY_PATH`. This is bad because:
+
+1. Running `nixos-rebuild` after a change to sane or it's configuration,
+   requires the user to logout and login, to see effects of the changes.  
+   
+2.
+[Apparently](https://github.com/NixOS/nixpkgs/issues/90201#issuecomment-683304279)
+`LD_LIBRARY_PATH` is cleared when an application runs with extra capabilities.
+This causes the current gnome-shell wayland wrapper to unset the
+`LD_LIBRARY_PATH` many scanner related programs now rely upon.
+
+Declarative wrappers should enable us to make such applications that use a
+`SANE_CONFIG_DIR` and `LD_LIBRARY_PATH` to be configured during the wrap phase,
+and get rid of these global environment variables.
+
+# Detailed design
+[design]: #detailed-design
+
+The end goal is to make the experience of getting a derivation wrapped as
+automatic as possible. A derivation that needs some environment variables in
+order to work will get these environment variables set in the wrapper by
+`mkDerivation`'s standard `fixupPhase`.  As a start, we'll (me probably) will
+introduce a new `makeWrapperAuto` setup hook that will take care of this in the
+way described as follows.
+
+As a start, we'll need to think about all packages in Nixpkgs that using them
+requires some environment variables to be set. Every such package will put in
+`$dev/nix-support/wrappers.json` a list of environment variables that are
+"linked" to this package. "linked" means that using this package requires these
+environment variables to be set in runtime.
+
+`makeWrapperAuto` will traverse all `buildInputs` and `propagatedBuildInputs`
+of a derivation, and look for a `wrappers.json` file in these inputs. It will
+collect all the environment variables that need to be set in the resulting
+executables, by merging all of the values of the environment variables in all
+of the inputs' `wrappers.json` files. `wrappers.json` might look like this for
+a given package:
+
+```json
+{
+  "GI_TYPELIB_PATH": [
+    "/nix/...gobject-introspection.../...",
+    "/nix/...librsvg.../...",
+  ],
+  "GIO_EXTRA_MODULES": [
+    "/nix/...dconf.../lib/gio/modules"
+  ],
+  "XDG_DATA_DIRS": [
+    "/nix/...gtk+3.../...",
+    "/nix/...gsettings-desktop-schemas.../..."
+  ],
```

Comment:
I thought about answers to some of these questions back when I implemented https://github.com/NixOS/nixpkgs/pull/85103/files#diff-aad93ba22adfe208c7b970a7b9d0f73dba9306b8a4fc4eaacfb2ac331aed9ea1R51-R74 . I prepared some answers in the RFC to be pushed, but I don't know:

- Why order would matter?
- I never encountered a case where paths should be escaped, or when other values should be escaped. In case of environment variables values that need escaping, they probably will not be originated from packages, and that can be handled in the `postInstall` snippet example below, by manually quoting values in the json file..

> We can also go further. Instead of setting `$PATH` to a massive list does it make more sense to build a folder of symlinks and set `$PATH` to a single element? If so will we special-case some variables or should we use a more detailed schema (as proposed below)?
> 

I thought about this exactly even back when I implemented https://github.com/NixOS/nixpkgs/pull/85103 so it will be possible to avoid what https://github.com/NixOS/nixpkgs/pull/84689 fixed. I addressed this idea in the changes I made to the RFC that I haven't pushed yet.

I'm not sure I understand:

> Furthermore does this replace or prepend or append? Replace seems the purest but commands such as `parallel -- command-from-my-path` are not uncommon. In that case there seems to be no good option.

Reply:
> Why order would matter?

If two packages have a `bin/foo` then the order will decide which one is used. For most of these `PATH`-like variables the first match is used.

> I never encountered a case where paths should be escaped, or when other values should be escaped.

It is rare but does happen. For things like `PATH` I don't think there is even a way to quote elements. You simply can't have a path that contains `:` in your `PATH`. At the very least it would be nice to raise an error. It would be even better if we can "alias" the path to a safe name or escape the invalid characters (if supported by the variable, which is rare)

> I'm not sure I understand:

My point is that the argument to a parallel is a command to run, but that command should be pulled from the PATH of the calling shell, not a PATH that may be set but a wrapper in the parallel derivation.

Reply:
> > Why order would matter?
> 
> If two packages have a `bin/foo` then the order will decide which one is used. For most of these `PATH`-like variables the first match is used.

I think it's super rare to have two packages providing the same binary like that, and finding them used together in the wrapping while wishing only one package's `out/bin` to rule..

Also paths escaping is too rare, so I think it'll be best to leave these topics out of scope for the RFC.

As for `parallel` commands. Do you mean `parallel` invocations in the builds of derivations? something like:

```nix
  preInstall = ''
    parallel -- commands that take a-long time
  '';
```

Reply:
I don't think collisions are super rare. After all nix has a priority system for some reason. I think we should at least define how this is handled. 

As for the escaping I think that it is fine to ignore this for now. But I think it should be mentioned explicitly. 

For the "parallel problem" I mean in general. Any time you have a wrapper that both needs commands itself as well as forwards a user's command we have this problem. I think we should discuss how this is handled. Currently in nixpkgs it is very inconsistent. 

But the most important point still stands. We need *some* configuration. If only to specify the separator. So I think that needs to be included.

Reply:
PATH order is support important. Tools like trash, safe-rm and molly-guard heavily rely on it. Also I commonly use a higher priority entry in PATH to create wrappers for none interactive shells. For environment variables which are similar to PATH like NIX_PATH order is also super important and ordering the entries for example alphabetically breaks things.

What is safe to do is deduplication when an entry is encountered the second time and I think this should be done by default to prevent to long environment variables.